File: matrix.py
# matrix
def print_matrix(m):
    for row in m:
        print(row)

def transpose(m):
    return [[m[j][i] for j in range(len(m))] for i in range(len(m[0]))]

def multiply(a, b):
    ra, ca = len(a), len(a[0])
    rb, cb = len(b), len(b[0])

    output = [[0 for _ in range(cb)] for _ in range(ra)]
    for i in range(ra):
        for j in range(cb):
            for k in range(rb):
                output[i][j] += a[i][k] * b[k][j]
    return output

# vector
def dot(a, b):
    sum = 0
    for i in range(len(a)):
        sum += a[i] * b[i]
    return sum

def norm(v):
    return (dot(v, v))**(1/2)

def right_multiply(m, v):
    output = []
    for row in m:
        output.append(dot(v, row))
    return output

def left_multiply(v, m):
    output = [0] * len(m[0])
    for i in range(len(m)):
        for j in range(len(m[0])):
            output[j] += v[i] * m[i][j]
    return output


if __name__ == "__main__":

    m = [[1,2],[3,4],[5,6]]
    v = [1,1]
    print_matrix(m)
    print_matrix(multiply(transpose(m), m))
    print(right_multiply(m, v))

File: c_matrix/matrix.c
#include <stdio.h>
#include <math.h>

void print_matrix(int row, int col, int m[row][col]) {
    for (int i = 0; i < row; i++) {
        for (int j = 0; j < col; j++) {
            printf("%d ", m[i][j]);
        }
        printf("\n");
    }
}

void transpose(int row, int col, int a[row][col], int b[col][row]) {
    for (int i = 0; i < row; i++) {
        for (int j = 0; j < col; j++) {
            b[j][i] = a[i][j];
        }
    }
}

void multiply(int r1, int c1, int c2, int a[r1][c1], int b[c1][c2], int c[r1][c2]) {
    for (int i = 0; i < r1; i++) {
        for (int j = 0; j < c2; j++) {
            c[i][j] = 0;
        }
    }

    for (int i = 0; i < r1; i++) {
        for (int j = 0; j < c2; j++) {
            for (int k = 0; k < c1; k++) {
                c[i][j] += a[i][k] * b[k][j];
            }
        }
    }
}

int dot(int l, int a[l], int b[l]) {
    int c = 0;
    for (int i = 0; i < l; i++) {
        c += a[i] * b[i];
    }
    return c;
}

float norm(int l, int v[l]) {
    return sqrt(dot(l, v, v));
}

void right_multiply(int row, int col, int m[row][col], int v[col], int c[row]) {
    for (int i = 0; i < row; i++) {
        c[i] = 0;
    }

    for (int i = 0; i < row; i++) {
        for (int j = 0; j < col; j++) {
            c[i] += m[i][j] * v[j];
        }
    }
}

void left_multiply(int row, int col, int v[row], int m[row][col], int c[col]) {
    for (int j = 0; j < col; j++) {
        c[j] = 0;
    }

    for (int i = 0; i < row; i++) {
        for (int j = 0; j < col; j++) {
            c[j] += v[i] * m[i][j];
        }
    }
}





int main() {
    int m[3][3] = {
    10, 23, 42,    
    1, 654, 0,  
    40652, 22, 0  
    };  
    print_matrix(3, 3, m);
    return 0;
}
